modules/sor/sorV2/lib/pools/weighted/weightedPool.ts

```diff
@@ -175,6 +188,83 @@ export class WeightedPool implements BasePool {
         return tokenInAmount;
     }
 
+    public addLiquiditySingleTokenExactIn(tokenIn: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        try {
+            // balances and amounts must be normalized to 1e18 fixed point - e.g. 1USDC => 1e18 not 1e6
+            const { tokenBalances, amountsIn, weights } = Array.from(this.tokenMap.values()).reduce(
+                (acc: { tokenBalances: bigint[]; amountsIn: bigint[]; weights: bigint[] }, weightedPoolToken) => {
+                    if (weightedPoolToken.token.address.toLowerCase() === tokenIn.address.toLowerCase()) {
+                        acc.amountsIn.push(amount.scale18);
+                    } else {
+                        acc.amountsIn.push(0n);
+                    }
+                    acc.tokenBalances.push(weightedPoolToken.scale18);
+                    acc.weights.push(weightedPoolToken.weight);
+                    return acc;
+                },
+                { tokenBalances: [], amountsIn: [], weights: [] },
+            );
+            const bptAmountOut = this._calcBptOutGivenExactTokensIn(
+                tokenBalances,
+                weights,
+                amountsIn,
+                this.totalShares,
+            );
+            return TokenAmount.fromRawAmount(bpt, bptAmountOut);
+        } catch (err) {
+            return TokenAmount.fromRawAmount(bpt, 0n);
+        }
+    }
+
+    public addLiquiditySingleTokenExactOut(tokenIn: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        try {
+            const token = this.tokenMap.get(tokenIn.wrapped);
+            if (!token) {
+                throw new Error(`Invalid Token: ${tokenIn.address}`);
+            }
+            const tokenInAmount = this._calcTokenInGivenExactBptOut(
+                token.scale18,
+                token.weight,
+                amount.scale18,
+                this.totalShares,
+            );
+            return TokenAmount.fromRawAmount(tokenIn, tokenInAmount);
+        } catch (err) {
+            return TokenAmount.fromRawAmount(tokenIn, 0n);
+        }
+    }
+
+    public removeLiquiditySingleTokenExactIn(tokenOut: Token, bpt: Token, amount: TokenAmount): TokenAmount {
+        const { amount: tokenBalance, weight: tokenWeight } = Array.from(this.tokenMap.values()).find(
+            (weightedPoolToken) => isSameAddress(weightedPoolToken.token.address, tokenOut.address),
```

Comment:
You should be able to use isSameAddress from within Token:
```suggestion
            (weightedPoolToken) => weightedPoolToken.token.isSameAddress(tokenOut.address),
```